apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/TransactionPool.cs

```diff
@@ -0,0 +1,247 @@
+        /// </summary>
+        /// <param name="session"></param>
+        /// <param name="client"></param>
+        public static Task PreWarmTransactionAsync(this SpannerClient client, Session session)
+        {
+            TransactionOptions options = session.GetLastUsedTransactionOptions();
+            var info = s_sessionInfoTable.GetOrAdd(session, s => new SessionInfo {SpannerClient = client});
+            return info.PreWarmAsync(session, options);
+        }
+
+        /// <summary>
+        /// Commits a transaction and internally marks the transaction as used.
+        /// </summary>
+        /// <param name="transaction"></param>
+        /// <param name="session"></param>
+        /// <param name="mutations"></param>
+        /// <returns></returns>
+        public static Task CommitAsync(this Transaction transaction, Session session, IEnumerable<Mutation> mutations)
+        {
+            return RunFinalMethodAsync(transaction, session,
+                info => info.SpannerClient.CommitAsync(session.SessionName, info.ActiveTransaction.Id, mutations));
+        }
+
+        /// <summary>
+        /// Rolls back a transaction and internally marks the transaction as used.
+        /// </summary>
+        /// <param name="transaction"></param>
+        /// <param name="session"></param>
+        /// <returns></returns>
+        public static Task RollbackAsync(this Transaction transaction, Session session)
+        {
+            return RunFinalMethodAsync(transaction, session,
+                info => info.SpannerClient.RollbackAsync(session.SessionName, info.ActiveTransaction.Id));
+        }
+
+        private static async Task RunFinalMethodAsync(Transaction transaction, Session session, Func<SessionInfo, Task> commitOrRollbackAction)
+        {
+            SessionInfo info;
+            if (s_sessionInfoTable.TryGetValue(session, out info))
+            {
+                //we should have already waited, but extra checking cannot hurt.
+                await info.WaitForPreWarmAsync().ConfigureAwait(false);
+                if (info.ActiveTransaction == null
+                    || info.ActiveTransaction.Id.IsEmpty
+                    || !Equals(info.ActiveTransaction.Id, transaction.Id))
+                {
+                    throw new InvalidOperationException("The transaction being committed was not found to have a valid entry.");
+                }
+
+                try
+                {
+                    await commitOrRollbackAction(info);
+                }
+                finally
+                {
+                    Session ignored;
+                    s_activeTransactionTable.TryRemove(info.ActiveTransaction.Id, out ignored);
+                }
+            }
+            else
+            {
+                throw new ArgumentException(
+                    "The given transaction was not found in the transaction pool. Only "
+                    +
```

Comment:
I'm fine either way, but isn't this just formatting?  Why would you say its not idiomatic in c#?  Been doing C# since before it was publicly released (>16 years) and I've never seen this enforced consistently.